nucypher/blockchain/eth/agents.py

```diff
@@ -456,16 +469,37 @@ def sample(self, quantity: int, duration: int, additional_ursulas: float = 1.7,
                 f'for {quantity} stakers we need a sample size of at least {n_select}.')
 
         system_random = random.SystemRandom()
-        n_tokens = self.contract.functions.getAllLockedTokens(duration).call()
+        n_tokens, stakers = self.contract.functions.getAllActiveStakers(duration).call()
         if n_tokens == 0:
             raise self.NotEnoughStakers('There are no locked tokens for duration {}.'.format(duration))
 
         for _ in range(attempts):
             points = sorted(system_random.randrange(n_tokens) for _ in range(n_select))
             self.log.debug(f"Sampling {n_select} stakers with random points: {points}")
 
-            addresses = set(self.contract.functions.sample(points, duration).call())
-            addresses.discard(str(BlockchainInterface.NULL_ADDRESS))
+            addresses = set()
+
+            previous_point = 0
+            point_index = 0
+            sum_of_locked_tokens = 0
+            staker_index = 0
+            stakers_len = len(stakers)
+            while staker_index < stakers_len and point_index < n_select:
+                current_staker = stakers[staker_index][0]
+                staker_tokens = stakers[staker_index][1]
+                if staker_tokens == 0:
+                    break
```

Comment:
why not throw an exception here?

Reply:
Zero means that no more stakers in an array, but result still could be more than `quantity`